content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -294,14 +294,289 @@ Unchecked Access
+
+Note that it is OK to call :ada:`Free` for an access value that has already
+been deallocated:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I : Integer_Access;
+
+    begin
+       I := new Integer;
+
+       Free (I);
+       Free (I);
+       Free (I);
+    end Show_Unchecked_Deallocation;
+
+The multiple calls to :ada:`Free` in this example don't cause any issues. In
+fact, calling :ada:`Free` to an access value that is equal to :ada:`null` is
+simply ignored.
+
+.. admonition:: In the Ada Reference Manual
+
+    - `4.8 Allocators <https://www.adaic.org/resources/add_content/standards/12rm/html/RM-4-8.html>`_
+    - `13.11.2 Unchecked Storage Deallocation <https://www.adaic.org/resources/add_content/standards/12rm/html/RM-13-11-2.html>`_
+
+
+Dereferencing after deallocation
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+An access value that points to a non-existent object is called a dangling
+reference. Of course, we shouldn't try to dereference a dangling reference
+because it causes the :ref:`access check <Adv_Ada_Access_Check>` to fail. For
+example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+    :class: ada-run-expect-failure
+
+    with Ada.Text_IO;   use Ada.Text_IO;
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I_1, I_2 : Integer_Access;
+
+    begin
+       I_1 := new Integer'(42);
+       I_2 := I_1;
+
+       Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       Put_Line ("I_2 = " & Integer'Image (I_1.all));
+
+       Free (I_1);
+
+       if I_1 /= null then
+          Put_Line ("I_1 = " & Integer'Image (I_1.all));
+       end if;
+
+       if I_2 /= null then
+          Put_Line ("I_2 = " & Integer'Image (I_1.all));
+       end if;
+    end Show_Unchecked_Deallocation;
+
+In this example, we allocate an object for :ada:`I_1` and make :ada:`I_2` point
+to the same object. Note, however, that we call :ada:`Free` for :ada:`I_1`, but
+not for :ada:`I_2`. Therefore, after that call to :ada:`Free`,
+:ada:`I_1 = null`, but :ada:`I_2` still points to the object originally
+allocated for :ada:`I_1`.
+
+As mentioned previously, when deallocating an object via a call to :ada:`Free`,
+the corresponding access value is set to :ada:`null`. In this case, we can
+check for :ada:`null` to detect a dangling reference and avoid dereferencing
+it.
+
+The strategy of checking for :ada:`null` before trying to dereferencing the
+access value works fine for :ada:`I_1`, but not for :ada:`I_2`, as we can see
+when running the example above. In the case of :ada:`I_2`, the access value is
+not :ada:`null`: :ada:`I_2` points to an object that doesn't exist anymore. As
+expected, dereferencing :ada:`I_2` raises the :ada:`Constraint_Error`
+exception.
```

Comment:
The evaluation of a dangling reference is erroneous, meaning the behavior in that case is undefined. If the compiler can detect the situation it can raise C_E, but more likely it will just give junk results for the dereference, or worse yet appear to work correctly if the junk is a valid value, or launch rockets, or anything else. This issue is discussed in RM 13.9.1 (13) and especially 13.11.2 (15).

Reply:
Good point! The code example was actually wrong. (This is now corrected.) Also, I rewrote this part to emphasize that the behavior is undefined.